File: packages/numlang-js/src/ordinal.ts
import { NumberToWordsOptions, defaultOptions } from "./options";
import { scales, ones, tens, toWords } from "./cardinal";

const ordinalSuffixes: { [key: number]: string } = {
  0: "th",
  1: "st",
  2: "nd",
  3: "rd",
  4: "th",
  5: "th",
  6: "th",
  7: "th",
  8: "th",
  9: "th",
};

const ordinalOnes: { [key: string]: string } = {
  one: "first",
  two: "second",
  three: "third",
  four: "fourth",
  five: "fifth",
  six: "sixth",
  seven: "seventh",
  eight: "eighth",
  nine: "ninth",
};

const ordinalTens: { [key: string]: string } = {
  ten: "tenth",
  twenty: "twentieth",
  thirty: "thirtieth",
  forty: "fortieth",
  fifty: "fiftieth",
  sixty: "sixtieth",
  seventy: "seventieth",
  eighty: "eightieth",
  ninety: "ninetieth",
};

function stringIncludes(str: string, substr: string): boolean {
  return str.indexOf(substr) !== -1;
}

function isFloat(n: number | string): boolean {
  if (typeof n === "number") return n % 1 !== 0;
  if (typeof n === "string")
    return n.includes(".") && !Number.isNaN(parseFloat(n));
  return false;
}

export function toOrdinal(number: number | string): string {
  let num: number;
  if (typeof number === "string") {
    if (isFloat(number)) {
      throw new Error(
        "Ordinal forms for floats are not standard in English. Please provide an integer.",
      );
    }
    num = parseInt(number, 10);
    if (isNaN(num)) throw new Error("Invalid number string provided");
  } else if (typeof number === "number") {
    if (!isFinite(number) || isNaN(number))
      throw new Error("Invalid number provided");
    if (isFloat(number)) {
      throw new Error(
        "Ordinal forms for floats are not standard in English. Please provide an integer.",
      );
    }
    num = number;
  } else {
    throw new Error(
      "Input must be a number or a string representation of a number",
    );
  }
  const absNum = Math.abs(num);
  if (absNum % 100 >= 11 && absNum % 100 <= 13) return num + "th";
  const lastDigit = absNum % 10;
  const suffix = ordinalSuffixes[lastDigit] || "th";
  return num + suffix;
}

export function toWordsOrdinal(
  number: number | string,
  options: NumberToWordsOptions = {},
): string {
  const mergedOptions = { ...defaultOptions, ...options };
  let num: number;
  if (typeof number === "string") {
    if (isFloat(number)) {
      throw new Error(
        "Ordinal word forms for floats are not standard in English. Please provide an integer.",
      );
    }
    num = parseInt(number, 10);
    if (isNaN(num)) throw new Error("Invalid number string provided");
  } else if (typeof number === "number") {
    if (!isFinite(number) || isNaN(number))
      throw new Error("Invalid number provided");
    if (isFloat(number)) {
      throw new Error(
        "Ordinal word forms for floats are not standard in English. Please provide an integer.",
      );
    }
    num = number;
  } else {
    throw new Error(
      "Input must be a number or a string representation of a number",
    );
  }
  if (num === 0) return "zeroth";
  let prefix = "";
  if (num < 0) {
    prefix = "negative ";
    num = Math.abs(num);
  }
  let words = toWords(num, mergedOptions);
  if (prefix) return prefix + toWordsOrdinal(Math.abs(num), mergedOptions);
  for (let i = 1; i < 10; i++) {
    if (words === ones[i]) return ordinalOnes[words];
  }
  for (const key in ordinalTens) {
    if (ordinalTens.hasOwnProperty(key) && words === key)
      return ordinalTens[key];
  }
  if (stringIncludes(words, "-")) {
    const parts = words.split("-");
    if (parts.length === 2) {
      const lastWord = parts[1];
      if (ordinalOnes[lastWord]) return parts[0] + "-" + ordinalOnes[lastWord];
    }
  }
  const lastSpace = words.lastIndexOf(" ");
  if (lastSpace !== -1) {
    const lastWord = words.substring(lastSpace + 1);
    for (let i = 0; i < scales.length; i++) {
      if (scales[i] && lastWord === scales[i])
        return words.substring(0, lastSpace + 1) + lastWord + "th";
    }
    if (ordinalOnes[lastWord])
      return words.substring(0, lastSpace + 1) + ordinalOnes[lastWord];
    if (ordinalTens[lastWord])
      return words.substring(0, lastSpace + 1) + ordinalTens[lastWord];
  }
  return words + "th";
}

File: packages/numlang-rs/src/ordinal.rs
use crate::cardinal::to_words;
use crate::Options;

const ORDINAL_SUFFIXES: [&str; 10] = ["th", "st", "nd", "rd", "th", "th", "th", "th", "th", "th"];

const ORDINAL_ONES: [(&str, &str); 9] = [
    ("one", "first"),
    ("two", "second"),
    ("three", "third"),
    ("four", "fourth"),
    ("five", "fifth"),
    ("six", "sixth"),
    ("seven", "seventh"),
    ("eight", "eighth"),
    ("nine", "ninth"),
];

const ORDINAL_TENS: [(&str, &str); 8] = [
    ("ten", "tenth"),
    ("twenty", "twentieth"),
    ("thirty", "thirtieth"),
    ("forty", "fortieth"),
    ("fifty", "fiftieth"),
    ("sixty", "sixtieth"),
    ("seventy", "seventieth"),
    ("eighty", "eightieth"),
    // ninety handled below
];

pub fn to_ordinal(number: impl Into<i64>) -> String {
    let num = number.into();
    if num < 0 {
        return format!("-{}", to_ordinal(-num));
    }
    let abs_num = num.abs();
    if abs_num % 100 >= 11 && abs_num % 100 <= 13 {
        format!("{}th", num)
    } else {
        let last_digit = (abs_num % 10) as usize;
        format!("{}{}", num, ORDINAL_SUFFIXES[last_digit])
    }
}

pub fn to_words_ordinal(number: impl Into<i64>, options: &Options) -> String {
    let num = number.into();
    if num == 0 {
        return "zeroth".to_string();
    }
    if num < 0 {
        return format!("negative {}", to_words_ordinal(-num, options));
    }
    let words = to_words(num as f64, options);

    // Check for simple cases
    for &(card, ord) in ORDINAL_ONES.iter() {
        if words == card {
            return format_with_options(ord, options);
        }
    }
    for &(card, ord) in ORDINAL_TENS.iter() {
        if words == card {
            return format_with_options(ord, options);
        }
    }
    if words == "ninety" {
        return format_with_options("ninetieth", options);
    }

    // Handle hyphenated (e.g., twenty-one)
    if let Some(idx) = words.find('-') {
        let (prefix, last) = words.split_at(idx + 1);
        let last = &last[0..];
        for &(card, ord) in ORDINAL_ONES.iter() {
            if last == card {
                return format_with_options(&format!("{}{}", prefix, ord), options);
            }
        }
    }

    // Handle last word (e.g., one thousand two hundred thirty-four)
    if let Some(idx) = words.rfind(' ') {
        let (prefix, last) = words.split_at(idx + 1);
        let last = &last[0..];
        for &(card, ord) in ORDINAL_ONES.iter() {
            if last == card {
                return format_with_options(&format!("{}{}", prefix, ord), options);
            }
        }
        for &(card, ord) in ORDINAL_TENS.iter() {
            if last == card {
                return format_with_options(&format!("{}{}", prefix, ord), options);
            }
        }
        if last == "ninety" {
            return format_with_options(&format!("{}ninetieth", prefix), options);
        }
    }

    // Default: add "th"
    format_with_options(&format!("{}th", words), options)
}

fn format_with_options(s: &str, options: &Options) -> String {
    let mut result = s.to_string();
    if options.append_only {
        result.push_str(" only");
    }
    if options.capitalize && !result.is_empty() {
        let mut chars = result.chars();
        if let Some(first) = chars.next() {
            result = first.to_uppercase().collect::<String>() + chars.as_str();
        }
    }
    if options.uppercase {
        result = result.to_uppercase();
    }
    result
}
